signac_dashboard/dashboard.py

```diff
@@ -582,17 +582,42 @@ def page_not_found(error):
 
         @dashboard.login_manager.unauthorized_handler
         def unauthorized_handler():
-            return self._render_error("Access token is required.")
+            request_url = request.url
+            if "module" not in request_url:
+                session["redirect_url"] = request_url
+            return render_template("login.html")
 
-        @dashboard.app.route("/login")
+        @dashboard.app.route("/login", methods=["GET", "POST"])
         def login():
-            provided_token = request.args.get("token")
+            redirect_url = session.pop("redirect_url", "/")
+            print("redirect url", redirect_url)
+            if flask_login.current_user.is_authenticated:
+                # in case the user goes to the login page via browser history
+                return redirect(redirect_url)
+
+            provided_token = request.args.get("token")  # None if not given
+            if request.method == "POST":
+                provided_token = request.form.get("token")
+
             if provided_token == self.config["ACCESS_TOKEN"]:
                 user = User(provided_token)
                 flask_login.login_user(user)
-                return redirect("/")
+                # logs user in and goes to page they were on
+                return redirect(redirect_url)
 
-            return self._render_error("Invalid token")
+            else:
+                if provided_token is None:
+                    # First time visiting page, so don't display error.
+                    return render_template("login.html")
+                flash("Incorrect token", "danger")
+                if request.method == "GET":
+                    return redirect("/login")
+                elif request.method == "POST":
+                    if redirect_url == "/":
+                        redirect_url = "/login"
+                    return redirect(redirect_url)
+                else:
```

Comment:
You pass `methods=["GET", "POST"]` argument to `route`. According to the docs, any other method will result in a method not allowed error instead of calling the defined route: https://werkzeug.palletsprojects.com/en/2.2.x/routing/#werkzeug.routing.Rule

Reply:
Thanks!